File: src/content/docs/topics/cs/coding-problems/arrays-and-hashing/242-valid-anagram-approach3.py
def is_anagram(s: str, t: str) -> bool:
    if len(s) != len(t):
        return False
    counts = [0] * 26
    for ch in s:
        counts[ord(ch) - ord('a')] += 1
    for ch in t:
        idx = ord(ch) - ord('a')
        counts[idx] -= 1
        if counts[idx] < 0:
            return False
    return True

assert is_anagram('anagram', 'nagaram') == True
assert is_anagram('rat', 'car') == False
assert is_anagram('a', 'a') == True
assert is_anagram('ab', 'ba') == True
assert is_anagram('ab', 'a') == False
assert is_anagram('', '') == True
print("all tests pass")

File: src/content/docs/topics/cs/coding-problems/arrays-and-hashing/242-valid-anagram-approach3.ts
function assert(condition: boolean, msg: string = ''): void {
    if (!condition) throw new Error(msg || 'Assertion failed');
}

function isAnagram(s: string, t: string): boolean {
    if (s.length !== t.length) return false;
    const counts = new Array(26).fill(0);
    for (const ch of s) counts[ch.charCodeAt(0) - 97]++;
    for (const ch of t) {
        const idx = ch.charCodeAt(0) - 97;
        counts[idx]--;
        if (counts[idx] < 0) return false;
    }
    return true;
}

assert(isAnagram('anagram', 'nagaram') === true);
assert(isAnagram('rat', 'car') === false);
assert(isAnagram('a', 'a') === true);
assert(isAnagram('ab', 'ba') === true);
assert(isAnagram('ab', 'a') === false);
assert(isAnagram('', '') === true);
console.log("all tests pass");
